File: src/multimedia/libqmediatimerange.zig
const C = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qmediatimerange.html
pub const qmediatimerange = struct {
    /// New constructs a new QMediaTimeRange object.
    ///
    ///
    pub fn New() ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_new();
    }

    /// New2 constructs a new QMediaTimeRange object.
    ///
    /// ``` start: i64, end: i64 ```
    pub fn New2(start: i64, end: i64) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_new2(@intCast(start), @intCast(end));
    }

    /// New3 constructs a new QMediaTimeRange object.
    ///
    /// ``` param1: ?*C.QMediaTimeRange__Interval ```
    pub fn New3(param1: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_new3(@ptrCast(param1));
    }

    /// New4 constructs a new QMediaTimeRange object.
    ///
    /// ``` range: ?*C.QMediaTimeRange ```
    pub fn New4(range: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_new4(@ptrCast(range));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange ```
    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        C.QMediaTimeRange_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#swap)
    ///
    /// ``` self: ?*C.QMediaTimeRange, other: ?*C.QMediaTimeRange ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QMediaTimeRange_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#detach)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn Detach(self: ?*anyopaque) void {
        C.QMediaTimeRange_Detach(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange__Interval ```
    pub fn OperatorAssignWithQMediaTimeRangeInterval(self: ?*anyopaque, param1: ?*anyopaque) void {
        C.QMediaTimeRange_OperatorAssignWithQMediaTimeRangeInterval(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#earliestTime)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn EarliestTime(self: ?*anyopaque) i64 {
        return C.QMediaTimeRange_EarliestTime(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#latestTime)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn LatestTime(self: ?*anyopaque) i64 {
        return C.QMediaTimeRange_LatestTime(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#intervals)
    ///
    /// ``` self: ?*C.QMediaTimeRange, allocator: std.mem.Allocator ```
    pub fn Intervals(self: ?*anyopaque, allocator: std.mem.Allocator) []?*C.QMediaTimeRange__Interval {
        const _arr: C.struct_libqt_list = C.QMediaTimeRange_Intervals(@ptrCast(self));
        defer C.libqt_free(_arr.data);
        const _ret = allocator.alloc(?*C.QMediaTimeRange__Interval, _arr.len) catch @panic("Memory allocation failed");
        const _data: [*]?*C.QMediaTimeRange__Interval = @ptrCast(@alignCast(_arr.data));
        for (0.._arr.len) |_i| {
            _ret[_i] = _data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#isEmpty)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn IsEmpty(self: ?*anyopaque) bool {
        return C.QMediaTimeRange_IsEmpty(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#isContinuous)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn IsContinuous(self: ?*anyopaque) bool {
        return C.QMediaTimeRange_IsContinuous(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#contains)
    ///
    /// ``` self: ?*C.QMediaTimeRange, time: i64 ```
    pub fn Contains(self: ?*anyopaque, time: i64) bool {
        return C.QMediaTimeRange_Contains(@ptrCast(self), @intCast(time));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#addInterval)
    ///
    /// ``` self: ?*C.QMediaTimeRange, start: i64, end: i64 ```
    pub fn AddInterval(self: ?*anyopaque, start: i64, end: i64) void {
        C.QMediaTimeRange_AddInterval(@ptrCast(self), @intCast(start), @intCast(end));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#addInterval)
    ///
    /// ``` self: ?*C.QMediaTimeRange, interval: ?*C.QMediaTimeRange__Interval ```
    pub fn AddIntervalWithInterval(self: ?*anyopaque, interval: ?*anyopaque) void {
        C.QMediaTimeRange_AddIntervalWithInterval(@ptrCast(self), @ptrCast(interval));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#addTimeRange)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange ```
    pub fn AddTimeRange(self: ?*anyopaque, param1: ?*anyopaque) void {
        C.QMediaTimeRange_AddTimeRange(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#removeInterval)
    ///
    /// ``` self: ?*C.QMediaTimeRange, start: i64, end: i64 ```
    pub fn RemoveInterval(self: ?*anyopaque, start: i64, end: i64) void {
        C.QMediaTimeRange_RemoveInterval(@ptrCast(self), @intCast(start), @intCast(end));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#removeInterval)
    ///
    /// ``` self: ?*C.QMediaTimeRange, interval: ?*C.QMediaTimeRange__Interval ```
    pub fn RemoveIntervalWithInterval(self: ?*anyopaque, interval: ?*anyopaque) void {
        C.QMediaTimeRange_RemoveIntervalWithInterval(@ptrCast(self), @ptrCast(interval));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#removeTimeRange)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange ```
    pub fn RemoveTimeRange(self: ?*anyopaque, param1: ?*anyopaque) void {
        C.QMediaTimeRange_RemoveTimeRange(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator+=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange ```
    pub fn OperatorPlusAssign(self: ?*anyopaque, param1: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_OperatorPlusAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator+=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange__Interval ```
    pub fn OperatorPlusAssignWithQMediaTimeRangeInterval(self: ?*anyopaque, param1: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_OperatorPlusAssignWithQMediaTimeRangeInterval(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator-=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange ```
    pub fn OperatorMinusAssign(self: ?*anyopaque, param1: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_OperatorMinusAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#operator-=)
    ///
    /// ``` self: ?*C.QMediaTimeRange, param1: ?*C.QMediaTimeRange__Interval ```
    pub fn OperatorMinusAssignWithQMediaTimeRangeInterval(self: ?*anyopaque, param1: ?*anyopaque) ?*C.QMediaTimeRange {
        return C.QMediaTimeRange_OperatorMinusAssignWithQMediaTimeRangeInterval(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange.html#clear)
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn Clear(self: ?*anyopaque) void {
        C.QMediaTimeRange_Clear(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QMediaTimeRange ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QMediaTimeRange_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qmediatimerange-interval.html
pub const qmediatimerange__interval = struct {
    /// New constructs a new QMediaTimeRange::Interval object.
    ///
    /// ``` other: ?*C.QMediaTimeRange__Interval ```
    pub fn New(other: ?*anyopaque) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_new(@ptrCast(other));
    }

    /// New2 constructs a new QMediaTimeRange::Interval object and invalidates the source QMediaTimeRange::Interval object.
    ///
    /// ``` other: ?*C.QMediaTimeRange__Interval ```
    pub fn New2(other: ?*anyopaque) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_new2(@ptrCast(other));
    }

    /// New3 constructs a new QMediaTimeRange::Interval object.
    ///
    ///
    pub fn New3() ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_new3();
    }

    /// New4 constructs a new QMediaTimeRange::Interval object.
    ///
    /// ``` start: i64, end: i64 ```
    pub fn New4(start: i64, end: i64) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_new4(@intCast(start), @intCast(end));
    }

    /// New5 constructs a new QMediaTimeRange::Interval object.
    ///
    /// ``` param1: ?*C.QMediaTimeRange__Interval ```
    pub fn New5(param1: ?*anyopaque) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_new5(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: ?*QMediaTimeRange__Interval, other: ?*QMediaTimeRange__Interval ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QMediaTimeRange__Interval_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: ?*QMediaTimeRange__Interval, other: ?*QMediaTimeRange__Interval ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QMediaTimeRange__Interval_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#start)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval ```
    pub fn Start(self: ?*anyopaque) i64 {
        return C.QMediaTimeRange__Interval_Start(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#end)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval ```
    pub fn End(self: ?*anyopaque) i64 {
        return C.QMediaTimeRange__Interval_End(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#contains)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval, time: i64 ```
    pub fn Contains(self: ?*anyopaque, time: i64) bool {
        return C.QMediaTimeRange__Interval_Contains(@ptrCast(self), @intCast(time));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#isNormal)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval ```
    pub fn IsNormal(self: ?*anyopaque) bool {
        return C.QMediaTimeRange__Interval_IsNormal(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#normalized)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval ```
    pub fn Normalized(self: ?*anyopaque) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_Normalized(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qmediatimerange__interval.html#translated)
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval, offset: i64 ```
    pub fn Translated(self: ?*anyopaque, offset: i64) ?*C.QMediaTimeRange__Interval {
        return C.QMediaTimeRange__Interval_Translated(@ptrCast(self), @intCast(offset));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QMediaTimeRange__Interval ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QMediaTimeRange__Interval_Delete(@ptrCast(self));
    }
};

File: src/multimedia/libqmediatimerange.cpp
#include <QList>
#include <QMediaTimeRange>
#define WORKAROUND_INNER_CLASS_DEFINITION_QMediaTimeRange__Interval
#include <qmediatimerange.h>
#include "libqmediatimerange.h"
#include "libqmediatimerange.hxx"

QMediaTimeRange* QMediaTimeRange_new() {
    return new QMediaTimeRange();
}

QMediaTimeRange* QMediaTimeRange_new2(long long start, long long end) {
    return new QMediaTimeRange(static_cast<qint64>(start), static_cast<qint64>(end));
}

QMediaTimeRange* QMediaTimeRange_new3(const QMediaTimeRange__Interval* param1) {
    return new QMediaTimeRange(*param1);
}

QMediaTimeRange* QMediaTimeRange_new4(const QMediaTimeRange* range) {
    return new QMediaTimeRange(*range);
}

void QMediaTimeRange_OperatorAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
    self->operator=(*param1);
}

void QMediaTimeRange_Swap(QMediaTimeRange* self, QMediaTimeRange* other) {
    self->swap(*other);
}

void QMediaTimeRange_Detach(QMediaTimeRange* self) {
    self->detach();
}

void QMediaTimeRange_OperatorAssignWithQMediaTimeRangeInterval(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
    self->operator=(*param1);
}

long long QMediaTimeRange_EarliestTime(const QMediaTimeRange* self) {
    return static_cast<long long>(self->earliestTime());
}

long long QMediaTimeRange_LatestTime(const QMediaTimeRange* self) {
    return static_cast<long long>(self->latestTime());
}

libqt_list /* of QMediaTimeRange__Interval* */ QMediaTimeRange_Intervals(const QMediaTimeRange* self) {
    QList<QMediaTimeRange::Interval> _ret = self->intervals();
    // Convert QList<> from C++ memory to manually-managed C memory
    QMediaTimeRange__Interval** _arr = static_cast<QMediaTimeRange__Interval**>(malloc(sizeof(QMediaTimeRange__Interval*) * _ret.length()));
    for (size_t i = 0; i < _ret.length(); ++i) {
        _arr[i] = new QMediaTimeRange::Interval(_ret[i]);
    }
    libqt_list _out;
    _out.len = _ret.length();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

bool QMediaTimeRange_IsEmpty(const QMediaTimeRange* self) {
    return self->isEmpty();
}

bool QMediaTimeRange_IsContinuous(const QMediaTimeRange* self) {
    return self->isContinuous();
}

bool QMediaTimeRange_Contains(const QMediaTimeRange* self, long long time) {
    return self->contains(static_cast<qint64>(time));
}

void QMediaTimeRange_AddInterval(QMediaTimeRange* self, long long start, long long end) {
    self->addInterval(static_cast<qint64>(start), static_cast<qint64>(end));
}

void QMediaTimeRange_AddIntervalWithInterval(QMediaTimeRange* self, const QMediaTimeRange__Interval* interval) {
    self->addInterval(*interval);
}

void QMediaTimeRange_AddTimeRange(QMediaTimeRange* self, const QMediaTimeRange* param1) {
    self->addTimeRange(*param1);
}

void QMediaTimeRange_RemoveInterval(QMediaTimeRange* self, long long start, long long end) {
    self->removeInterval(static_cast<qint64>(start), static_cast<qint64>(end));
}

void QMediaTimeRange_RemoveIntervalWithInterval(QMediaTimeRange* self, const QMediaTimeRange__Interval* interval) {
    self->removeInterval(*interval);
}

void QMediaTimeRange_RemoveTimeRange(QMediaTimeRange* self, const QMediaTimeRange* param1) {
    self->removeTimeRange(*param1);
}

QMediaTimeRange* QMediaTimeRange_OperatorPlusAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
    QMediaTimeRange& _ret = self->operator+=(*param1);
    // Cast returned reference into pointer
    return &_ret;
}

QMediaTimeRange* QMediaTimeRange_OperatorPlusAssignWithQMediaTimeRangeInterval(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
    QMediaTimeRange& _ret = self->operator+=(*param1);
    // Cast returned reference into pointer
    return &_ret;
}

QMediaTimeRange* QMediaTimeRange_OperatorMinusAssign(QMediaTimeRange* self, const QMediaTimeRange* param1) {
    QMediaTimeRange& _ret = self->operator-=(*param1);
    // Cast returned reference into pointer
    return &_ret;
}

QMediaTimeRange* QMediaTimeRange_OperatorMinusAssignWithQMediaTimeRangeInterval(QMediaTimeRange* self, const QMediaTimeRange__Interval* param1) {
    QMediaTimeRange& _ret = self->operator-=(*param1);
    // Cast returned reference into pointer
    return &_ret;
}

void QMediaTimeRange_Clear(QMediaTimeRange* self) {
    self->clear();
}

void QMediaTimeRange_Delete(QMediaTimeRange* self) {
    delete self;
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new(const QMediaTimeRange__Interval* other) {
    return new QMediaTimeRange::Interval(*other);
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new2(QMediaTimeRange__Interval* other) {
    return new QMediaTimeRange::Interval(std::move(*other));
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new3() {
    return new QMediaTimeRange::Interval();
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new4(long long start, long long end) {
    return new QMediaTimeRange::Interval(static_cast<qint64>(start), static_cast<qint64>(end));
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_new5(const QMediaTimeRange__Interval* param1) {
    return new QMediaTimeRange::Interval(*param1);
}

void QMediaTimeRange__Interval_CopyAssign(QMediaTimeRange__Interval* self, QMediaTimeRange__Interval* other) {
    *self = *other;
}

void QMediaTimeRange__Interval_MoveAssign(QMediaTimeRange__Interval* self, QMediaTimeRange__Interval* other) {
    *self = std::move(*other);
}

long long QMediaTimeRange__Interval_Start(const QMediaTimeRange__Interval* self) {
    return static_cast<long long>(self->start());
}

long long QMediaTimeRange__Interval_End(const QMediaTimeRange__Interval* self) {
    return static_cast<long long>(self->end());
}

bool QMediaTimeRange__Interval_Contains(const QMediaTimeRange__Interval* self, long long time) {
    return self->contains(static_cast<qint64>(time));
}

bool QMediaTimeRange__Interval_IsNormal(const QMediaTimeRange__Interval* self) {
    return self->isNormal();
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_Normalized(const QMediaTimeRange__Interval* self) {
    return new QMediaTimeRange::Interval(self->normalized());
}

QMediaTimeRange__Interval* QMediaTimeRange__Interval_Translated(const QMediaTimeRange__Interval* self, long long offset) {
    return new QMediaTimeRange::Interval(self->translated(static_cast<qint64>(offset)));
}

void QMediaTimeRange__Interval_Delete(QMediaTimeRange__Interval* self) {
    delete self;
}
